dask/layers.py

```diff
@@ -178,16 +202,30 @@ def keys(*args):
                 result = [keys(*(args + (i,))) for i in range(numblocks[ind])]
             return result
 
-        self._cached_keys = result = keys()
+        self._cached_input_keys = result = keys()
         return result
 
+    def _keys_from_overlap_operation(self):
+        if self._interior_keys is not None and self._expand_keys_func is not None:
```

Comment:
I don't actually see where the cache of `_interior_keys` and `self._expand_keys_func` happens. Aside from being ugly, this pattern seems a risky if there is any part of the layer that is mutable. 

If the layer is fully immutable, then it might make sense to use properties instead of methods like we do in array/core.py

https://github.com/dask/dask/blob/e4799c0498b5e5877705bb5542d8d01116ee1320/dask/array/core.py#L1336-L1338